### PR TITLE
apps/testcase/arastorage : add scenario TCs for arastorage_itc

### PR DESCRIPTION
**Pull request by arvinmittal**
file path: TizenRT/apps/examples/testcase/ta_tc/arastorage/itc
file name: itc_arastorage_main.c
add scenario:
itc_arastorage_db_init_deinit_multi_time_p
itc_arastorage_db_print_header_after_deinit_p
itc_arastorage_db_print_tuple_after_deinit_p
itc_arastorage_db_print_value_after_deinit_p
itc_arastorage_cursor_get_count_after_deinit_p
itc_arastorage_cursor_get_row_after_deinit_p
itc_arastorage_db_query_create_remove_relation_p
itc_arastorage_db_query_recreate_relation_p
itc_arastorage_db_query_reremove_relation_p
itc_arastorage_db_query_create_remove_multi_relation_p
itc_arastorage_db_query_create_remove_relation_multiple_time_p
itc_arastorage_db_get_result_message_all_p
itc_arastorage_db_get_result_message_n
itc_arastorage_db_print_header_n
itc_arastorage_db_print_tuple_n
itc_arastorage_db_print_value_n
itc_arastorage_cursor_get_count_n
itc_arastorage_cursor_move_first_n
itc_arastorage_cursor_move_next_n
itc_arastorage_cursor_move_prev_n
itc_arastorage_cursor_move_last_n
itc_arastorage_cursor_move_to_n
itc_arastorage_cursor_is_first_row_n
itc_arastorage_cursor_is_last_row_n
itc_arastorage_cursor_get_attr_type_n
itc_arastorage_cursor_get_attr_name_n
itc_arastorage_cursor_get_attr_index_n
itc_arastorage_cursor_get_int_value_n
itc_arastorage_cursor_get_long_value_n
itc_arastorage_cursor_get_row_p
itc_arastorage_cursor_get_row_n
itc_arastorage_cursor_get_double_value_n
itc_arastorage_cursor_get_string_value_n
itc_arastorage_db_cursor_free_p
itc_arastorage_db_cursor_free_n

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /****************************************************************************
 *
-* Copyright 2016 Samsung Electronics All Rights Reserved.
+* Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: Don't modify this. Let's revet.

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 2]
```diff
@@ -77,13 +71,410 @@ char g_query[QUERY_LENGTH];
...
 
+/**
+* @testcase             db_cleanup
+* @brief                To de-initialize database resources and to remove relation name from DB
+* @scenario             If used api returns DB_OK
+* @apicovered           db_deinit
+* @precondition         none
+* @postcondition        none
+*/
+void db_cleanup(const char *rel)
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: Let's add "static" for all of local functions.

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 3]
```diff
@@ -639,38 +1658,132 @@ void itc_arastorage_cursor_get_row_tc_p(void)
...
+	TC_SUCCESS_RESULT();
+}
+
 int itc_arastorage_launcher(int argc, FAR char *argv[])
 {
 	total_fail = 0;
 	total_pass = 0;
 	itc_arastorage_db_init_deinit_p();
+	itc_arastorage_db_init_deinit_multi_time_p();
+	itc_arastorage_db_print_header_after_deinit_p();
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: xx_after_deinit functions are p? not n?

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 4]
```diff
@@ -77,13 +73,425 @@ char g_query[QUERY_LENGTH];
...
+	g_cursor = db_query(g_query);
+	TC_ASSERT_NEQ_CLEANUP("db_query", g_cursor, NULL, db_cleanup(rel_name));
+
+	ret = db_print_value(g_cursor, 2);
+
+	ret = db_deinit();
+	TC_ASSERT_EQ_CLEANUP("db_deinit", ret, DB_OK, db_cleanup(rel_name));
+
+	ret = db_print_value(g_cursor, 2);
+	TC_ASSERT_EQ_CLEANUP("db_print_value", ret, DB_CURSOR_ERROR, db_cleanup(rel_name));
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: http://suprem.sec.samsung.net/jira/browse/WCTIAPP-2562

Comment (Taejun-Kwon) [reply apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: please check my comment.

**Comment by seinfra** [comment]
Code Rule Check Result:
apps/examples/testcase/tc_common.h:37: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:68: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:80: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:92: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:104: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:116: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:128: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by sunghan-chang** [comment]
@arvinmittal Same as previous comment on other pull request. Please remove unnecessary commits.

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 5]
```diff
@@ -77,13 +73,425 @@ char g_query[QUERY_LENGTH];
...
+* @brief                To check initialization and deinitialization database resources
+* @scenario             check iterative initialization and deinitialization database operation returns DB_OK
+* @apicovered           db_deinit, db_init
+* @precondition         none
+* @postcondition        none
+*/
+void itc_arastorage_db_init_deinit_multi_time_p(void)
+{
+	db_result_t ret;
+	int iteration = 0;
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: Why you use useless variable here?
int i;
for (i = 0; i < 10; ++i)

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 6]
```diff
@@ -77,13 +73,425 @@ char g_query[QUERY_LENGTH];
...
+* @brief                To check print of the related information : relation, attribute name
+* @scenario             If used api returns DB_OK
+* @apicovered           db_print_header , db_query
+* @precondition         Database resource must be deinitialized
+* @postcondition        none
+*/
+void itc_arastorage_db_print_header_after_deinit_n(void)
+{
+	db_result_t ret;
+	const char *rel_name = "REL_1";
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: why don't you use RELATION_NAME?

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 7]
```diff
@@ -77,13 +73,425 @@ char g_query[QUERY_LENGTH];
...
+* @testcase             db_cleanup
+* @brief                To de-initialize database resources and to remove relation name from DB
+* @scenario             If used api returns DB_OK
+* @apicovered           db_exec
+* @precondition         none
+* @postcondition        none
+*/
+static void db_cleanup(const char *rel)
+{
+	char query[QUERY_LENGTH];
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: you already defined global variable for query, g_query.

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 8]
```diff
@@ -77,13 +73,425 @@ char g_query[QUERY_LENGTH];
...
+	g_cursor = NULL;
+	g_cursor = db_query(g_query);
+	TC_ASSERT_NEQ_CLEANUP("db_query", g_cursor, NULL, db_cleanup(rel_name));
+
+	ret = db_deinit();
+	TC_ASSERT_EQ_CLEANUP("db_deinit", ret, DB_OK, db_cleanup(rel_name));
+
+	cursor_row_t count;
+
+	count = cursor_get_count(g_cursor);
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: count is useless here, use function call directly.

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 9]
```diff
@@ -77,13 +73,425 @@ char g_query[QUERY_LENGTH];
...
+* @brief            Get current row in cursor after deinit
+* @scenario         Get current row in cursor after db deinit
+* @apicovered       cursor_get_row
+* @precondition     Database resource must be initialized
+* @postcondition    none
+*/
+void itc_arastorage_cursor_get_row_after_deinit_n(void)
+{
+	db_result_t ret;
+	const char *rel_name = "REL_COUNT";
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: I checked some case, all case call clean up at the last of tc.
then you don't have to define each of relation name..

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 10]
```diff
@@ -127,57 +535,184 @@ void itc_arastorage_startup_p(void)
...
+* @postcondition        none
 */
-void itc_arastorage_cleanup_p(void)
+void itc_arastorage_db_query_create_remove_relation_p(void)
 {
-	if (g_init) {
-		db_deinit();
+	TC_ASSERT_EQ("ARASTORAGE_STARTUP", g_check, true);
+	db_result_t ret;
+	char query[QUERY_LENGTH];
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: g_query

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 11]
```diff
@@ -187,34 +722,116 @@ void itc_arastorage_db_init_deinit_p(void)
...
+* @apicovered           db_get_result_message
+* @precondition         none
+* @postcondition        none
+*/
+void itc_arastorage_db_get_result_message_all_p(void)
+{
+	TC_ASSERT_EQ("ARASTORAGE_STARTUP", g_check, true);
+	int count;
+	int index = 0;
+	db_result_t errors[] = { DB_FINISHED, DB_GOT_ROW, DB_OK, DB_LIMIT_ERROR,
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: struct db_result_entry {
   db_result_t code;
   char *msg;
}
for (index = 0; ;;) {
    printf("Error code%d\n", entry[index].code);
    TC_ASSERT_EQ("db_get_result_message", strcmp(db_get_result_message(entry[index].code), entry[index].msg)), 0);
}

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 12]
```diff
@@ -187,34 +722,116 @@ void itc_arastorage_db_init_deinit_p(void)
...
-	TC_ASSERT_NEQ("db_get_result_message : DB_IMPLEMENTATION_ERROR", ret, 0);
-	TC_ASSERT_NEQ("db_get_result_message : DB_INDEX_ERROR", ret, 0);
-	TC_ASSERT_NEQ("db_get_result_message : DB_NAME_ERROR", ret, 0);
-	TC_ASSERT_NEQ("db_get_result_message : DB_BUSY_ERROR", ret, 0);
-	TC_ASSERT_NEQ("db_get_result_message : DB_INCONSISTENCY_ERROR", ret, 0);
-	TC_ASSERT_NEQ("db_get_result_message : DB_ARGUMENT_ERROR", ret, 0);
-	TC_ASSERT_NEQ("db_get_result_message : DB_FULL_ERROR", ret, 0);
-	TC_ASSERT_NEQ("db_get_result_message : DB_CURSOR_ERROR", ret, 0);
-	TC_ASSERT_NEQ("db_get_result_message : DB_EMPTY_CURSOR_ERROR", ret, 0);
+	TC_ASSERT_EQ("db_get_result_message", strcmp(DB_UNKNOWN_ERROR_MSG, db_get_result_message(num)), 0);
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: use 200 directly.

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c [hunk 13]
```diff
@@ -223,19 +840,41 @@ void itc_arastorage_db_get_result_message_p(void)
...
+* @testcase         itc_arastorage_db_print_header_n
+* @brief            Print a header of query result with invalid argument
+* @scenario         Print relation, attribute name of query result with NULL value
+* @apicovered       db_print_header
+* @precondition     none
+* @postcondition    none
+*/
+static void itc_arastorage_db_print_header_n(void)
+{
+	TC_ASSERT_EQ("ARASTORAGE_STARTUP", g_check, true);
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: what is g_check??

**Comment by Taejun-Kwon** [comment]
It is very hard to review hundreds line at once...
please apply step by step from now....

File: apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c (hunk 13 above)
Comment (arvinmittal) [reply apps/examples/testcase/ta_tc/arastorage/itc/itc_arastorage_main.c]: g_check is global flag used in "itc_arastorage_startup_p".  It is marked as false on arastorage_startup failure. Arastorage_startup needs to run successfully before running other ITCs.  Inside those ITCs , g_check validates the successful completion of Arastorage_startup function.

**Comment by arvinmittal** [comment]
Resolved merge conflicts and fix review comments in new commit. Also done ITCs name change.
new commit PR : https://github.com/Samsung/TizenRT/pull/1072

**Pull request closed without merging**